### PR TITLE
PAYE connector and service layer

### PR DESCRIPTION
**Pull request by chrisjameswright**

### CODE REVIEW

File: test/services/PayeAtsServiceSpec.scala [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import scala.concurrent.Future
+import scala.io.Source
+
+class PayeAtsServiceSpec extends UnitSpec with MockitoSugar with JsonUtil with GuiceOneAppPerTest with ScalaFutures with IntegrationPatience {
+
+  implicit val hc = HeaderCarrier()
+  val expectedResponse: JsValue = readJson("/paye_ats.json")
+  private val currentYear = 2018
+
+  private def readJson(path: String) = {
```
Comment (danewatts) [inline test/services/PayeAtsServiceSpec.scala]: Im a little confused why we have both this function and the JsonUtil trait inherited. Is this function just for the purpose of loading it as a Stream? If so, can we remove the JsonUtil. If we dont care about it being a stream, is there a reason why we arent using the functions in JsonUtil?

File: test/services/PayeAtsServiceSpec.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+      val result = getPayeATSData(testNino, currentYear).futureValue
+
+      result shouldBe Right(expectedResponse.as[PayeAtsData])
+    }
+
+    "return a INTERNAL_SERVER_ERROR response after receiving JsResultException while json parsing" in new TestService {
+
+      when(middleConnector.connectToPayeATS(eqTo(testNino), eqTo(currentYear))(any[HeaderCarrier]))
+        .thenReturn(Future.failed(new JsResultException(List())))
```
Comment (danewatts) [inline test/services/PayeAtsServiceSpec.scala]: I get this is to test the `.as[PayeAtsData]` part, but this mock is suggesting that middleConnector is throwing this exception. If we changed this line to be a `validate.fold`, this test case would still pass (and probably be missed for removal) even though it shouldn't happen. I am not particularly fussed about this, but please bear this in mind.

File: app/services/PayeAtsService.scala [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  def getPayeATSData(nino: Nino, taxYear: Int)(implicit hc: HeaderCarrier):Future[Either[HttpResponse,PayeAtsData]] = {
+     middleConnector.connectToPayeATS(nino,taxYear) map { response =>
+       response status match {
+         case OK =>  Right(response.json.as[PayeAtsData])
+         case _ => Left(response)
+       }
+     } recover {
+       case e: BadRequestException => Left(HttpResponse(BAD_REQUEST))
+       case e: NotFoundException   => Left(HttpResponse(NOT_FOUND))
```
Comment (danewatts) [inline app/services/PayeAtsService.scala]: Is there a reason that the flow for this call is:
If we get a BadRequest, the connector turns it from HttpResponse to exception, which then gets changed back into the HttpResponse? Is it to omit the response body/headers?

**Comment by chrisjameswright** [comment]
Amended the one comment we agreed should be addressed @danewatts

**Pull request merged**